### PR TITLE
[@types/react-redux] add types for hook factories

### PR DESCRIPTION
**Pull request by DylanVann**
Updates type definitions for changes in [this PR](https://github.com/reduxjs/react-redux/pull/1309) to `react-redux`.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/reduxjs/react-redux/pull/1309
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Review by octylFractal — APPROVED** [review]
Looks good.

I think this should work already, but I assume you can do:
```ts
const myUseSelector = createSelectorHook();
const useTypedSelector: TypedUseSelectorHook<RootState> = myUserSelector;
```
like you can with the default hooks? Might be worth adding a test for it.

**Comment by DylanVann** [comment]
@kenzierocks I didn't know about that functionality. Neat. I added a test.

File: types/react-redux/index.d.ts [hunk 1]
```diff
@@ -578,4 +579,28 @@ export interface TypedUseSelectorHook<TState> {
  */
 export function useStore<S = any, A extends Action = AnyAction>(): Store<S, A>;
 
+/**
+ * Hook factory, which creates a `useSelector` hook bound to a given context.
+ *
+ * @param Context passed to your `<Provider>`.
+ * @returns A `useSelector` hook bound to the specified context.
+ */
+export function createSelectorHook(context?: Context<ReactReduxContextValue>): typeof useSelector;
```
Comment (joshjg) [inline types/react-redux/index.d.ts]: Based on the official docs:
```
const MyContext = React.createContext(null)

// Export your custom hooks if you wish to use them in other files.
export const useStore = createStoreHook(MyContext)
export const useDispatch = createDispatchHook(MyContext)
export const useSelector = createSelectorHook(MyContext)
```

Shouldn't the type for the context argument be `Context<ReactReduxContextValue | null>`? I can create a PR if so.

Comment (abirmingham) [reply types/react-redux/index.d.ts]: It certainly seems so. The current typings appear to be incompatible with the docs recommendation - at least I was unable to achieve the docs recommendation with the current typings.

**Pull request merged**